planner/core/rule_join_reorder.go

```diff
@@ -31,25 +31,43 @@ import (
 // construct a new join order based on a reorder algorithm.
 //
 // For example: "InnerJoin(InnerJoin(a, b), LeftJoin(c, d))"
-// results in a join group {a, b, LeftJoin(c, d)}.
-func extractJoinGroup(p LogicalPlan) (group []LogicalPlan, eqEdges []*expression.ScalarFunction, otherConds []expression.Expression) {
+// results in a join group {a, b, c, d}.
+func extractJoinGroup(p LogicalPlan) (group []LogicalPlan, eqEdges []*expression.ScalarFunction,
+	otherConds []expression.Expression, joinTypes []JoinType) {
 	join, isJoin := p.(*LogicalJoin)
-	if !isJoin || join.preferJoinType > uint(0) || join.JoinType != InnerJoin || join.StraightJoin {
-		return []LogicalPlan{p}, nil, nil
+	if !isJoin || join.preferJoinType > uint(0) || join.StraightJoin ||
+		(join.JoinType != InnerJoin && join.JoinType != LeftOuterJoin && join.JoinType != RightOuterJoin) ||
+		((join.JoinType == LeftOuterJoin || join.JoinType == RightOuterJoin) && join.EqualConditions == nil) {
```

Comment:
Are there any test cases for this?
When we have a `t1 join t2 left join t3 join t4`, but there are no equal conditions for `t2 left join t3`. Can we consider the `{t1, t2, t4} left join t3`?

Reply:
If t2 left join t3 has no equal condition then eqConds is nill, we will not split left join, there will be {t1, t2 left join t3 join t4}